### PR TITLE
Allow using non-default django cache

### PR DESCRIPTION
**Pull request by induane**
It is handy to be able to flush the ldap cache without flushing the entire default django cache. That can reset non-token based logins and all kinds of data you may or may not want. Being able to configure it to use it's own independent cache within the Django caching framework is reallllly handy!

### CODE REVIEW

File: django_auth_ldap/backend.py [hunk 1]
```diff
@@ -112,6 +112,15 @@ def __getstate__(self):
             k: v for k, v in self.__dict__.items() if k not in ["_settings", "_ldap"]
         }
 
+    @property
+    def cache(self):
+        try:
+            return self._cache
+        except AttributeError:
+            cache_name = getattr(self.settings, "AUTH_LDAP_CACHE", "default")
+            self._cache = caches[cache_name]
```
Comment (jdufresne) [inline django_auth_ldap/backend.py:121]: Is there some performance reason the cache the lookup? Or just habit.

Either way, this is re-implementing Django's `cached_property` co can use that instead.

File: django_auth_ldap/backend.py [hunk 2]
```diff
@@ -359,10 +368,7 @@ def authenticate(self, password):
             logger.debug("Authentication failed for {}: {}".format(self._username, e))
         except ldap.LDAPError as e:
             results = ldap_error.send(
-                type(self.backend),
-                context="authenticate",
-                user=self._user,
-                exception=e,
+                type(self.backend), context="authenticate", user=self._user, exception=e
```
Comment (jdufresne) [inline django_auth_ldap/backend.py:371]: Please revert unrelated line changes.

File: django_auth_ldap/backend.py (hunk 1 above)
Comment (induane) [reply django_auth_ldap/backend.py:121]: Saves an attribute lookup when we hide it behind the property - but yea - forgot about the cached property decorator. I'll update htat.

File: django_auth_ldap/backend.py (hunk 2 above)
Comment (induane) [reply django_auth_ldap/backend.py:371]: Ahh - can do - the black pass did that.

**Comment by induane** [comment]
@jdufresne  I'm not sure what to do test wise - it appears to be failing because of something with the black formatter.

File: django_auth_ldap/backend.py [hunk 3]
```diff
@@ -112,6 +113,10 @@ def __getstate__(self):
             k: v for k, v in self.__dict__.items() if k not in ["_settings", "_ldap"]
         }
 
+    @cached_property
+    def cache(self):
+        return caches[getattr(self.settings, "AUTH_LDAP_CACHE", "default")]
+
```
Comment (francoisfreitag) [inline django_auth_ldap/backend.py:119]: This attribute is unused.

File: docs/reference.rst [hunk 4]
```diff
@@ -354,6 +354,33 @@ in the directory. The filter parameter should contain the placeholder
...
 
+.. setting:: AUTH_LDAP_CACHE
+
+AUTH_LDAP_CACHE
+~~~~~~~~~~~~~~~
+
+Default: ``default``
+
+The Django cache framework can store and manage multiple caches, each with
+their own unique name. By default the default cache is used. Setting this value
```
Comment (francoisfreitag) [inline docs/reference.rst:365]:
```suggestion
their own unique name. By default, the default cache is used. Setting this value
```

File: tests/tests.py [hunk 5]
```diff
@@ -150,7 +150,8 @@ def tearDownClass(cls):
 
     def setUp(self):
         super().setUp()
-        cache.clear()
+        for cache in caches.values():
```
Comment (francoisfreitag) [inline tests/tests.py:153]:
```suggestion
        for cache in caches.all():
```

File: tests/tests.py [hunk 6]
```diff
@@ -35,7 +35,7 @@
 import slapdtest
 from django.contrib.auth import authenticate, get_backends
 from django.contrib.auth.models import Group, Permission, User
-from django.core.cache import cache
+from django.core.cache import caches
```
Comment (francoisfreitag) [inline tests/tests.py:38]:
```suggestion
from django.core.cache import cache, caches
```

**Review by francoisfreitag — COMMENTED** [review]
That’s a nice improvement!

To test the improvement, can:
1. trigger an operation that creates a cache entry,
2. verify:
  1. that the cache entry was created in the configured cache, and
  2. that the cache entry does not exist Django default cache

**Comment by induane** [comment]
Left this in a bit of a mess and am now just going to rebase on current master.
